File: 77.组合/combine.go
package problem0077

func combine(n int, k int) [][]int {
	var nums []int
	var result [][]int
	for i := 1; i <= n; i++ {
		nums = append(nums, i)
	}
	backTracking(nums, k, &result, []int{}, 0)
	return result
}

func backTracking(nums []int, k int, result *[][]int, temp []int, index int) bool {
	if len(temp) == k {
		t := make([]int, len(temp))
		copy(t, temp)
		*result = append(*result, t)
		return true
	} else if len(temp) > k {
		return false
	}
	for i, l := index, len(nums); i < l; i++ {
		temp = append(temp, nums[i])
		if !backTracking(nums, k, result, temp, i+1) {
			break
		}
		temp = temp[:len(temp)-1]
	}
	return true
}

File: 77.组合/combine.js
function combine(n, k) {
	const result = []
	const used = new Array(n).fill(false)

	const backTrack = (combine) => {
		if (combine.length === k) {
			result.push(combine)
			return
		}
		for (let i = 1; i <= n; i++) {
			if (used[i - 1]) {
				continue
			}
			// [4, 1] 和 [1, 4]被视为相同，因此如果 i < combine最后一位，continue
			if (combine.length && i < combine[combine.length - 1]) {
				continue
			}
			used[i - 1] = true
			combine.push(i)
			backTrack([...combine])
			combine.pop()
			used[i - 1] = false
		}
	}
	backTrack([])
	return result
}